2003-04-28  Anton Kachalov <mouse@altlinux.org>

	* apt-pkg/versionmatch.cc (pkgVersionMatch::MatchVer):
	Ignore serial while matching version string.
	Check both for version and version-release while matching version string.

--- apt-0.5.5cnc4.1/apt-pkg/versionmatch.cc.orig	2003-03-25 19:40:45 +0300
+++ apt-0.5.5cnc4.1/apt-pkg/versionmatch.cc	2003-04-28 19:35:30 +0400
@@ -124,18 +124,36 @@ pkgVersionMatch::pkgVersionMatch(string 
 /* */
 bool pkgVersionMatch::MatchVer(const char *A,string B,bool Prefix)
 {   
-   const char *Ab = A;
-   const char *Ae = Ab + strlen(A);
+   string s(A), sc(A);
+   const char *Ab = s.c_str(), *Ac = sc.c_str();
+
+   for (string::iterator i = s.begin(), k = sc.begin(); i != s.end(); ++i,++k)
+   {
+      if (*i == ':')
+      {
+         Ab = &(*i) + 1;
+	 Ac = &(*k) + 1;
+      }
+      else if (*i == '-')
+      {
+         *i = 0;
+	 break;
+      }
+   }
+
+   const char *Ae = Ab + strlen(Ab);
+   const char *Af = Ac + strlen(Ac);
    
    // Strings are not a compatible size.
-   if ((unsigned)(Ae - Ab) != B.length() && Prefix == false ||
-       (unsigned)(Ae - Ab) < B.length())
-      return false;
-   
-   // Match (leading?)
-   if (stringcasecmp(B,Ab,Ab + B.length()) == 0)
-      return true;
-   
+   if (((unsigned)(Ae - Ab) == B.length() || Prefix == true) &&
+       (unsigned)(Ae - Ab) >= B.length() &&
+       stringcasecmp(B,Ab,Ab + B.length()) == 0)
+       return true;
+   else if (((unsigned)(Af - Ac) == B.length() || Prefix == true) &&
+       (unsigned)(Af - Ac) >= B.length() &&
+       stringcasecmp(B,Ac,Ac + B.length()) == 0)
+       return true;
+
    return false;
 }
 									/*}}}*/
